File: sdk/go/compact.go
package microsandbox

import (
	"context"
	"encoding/json"

	"github.com/superradcompany/microsandbox/sdk/go/internal/ffi"
)

// DiskCompactionOptions selects explicit maintenance, never persisted desired configuration.
type DiskCompactionOptions struct {
	// Layers limits the oldest sealed physical layers per disk including the base, excluding the writable head.
	// Nil selects all sealed layers; the limit must be at least two. Chains with fewer than two sealed layers are skipped.
	Layers *uint32 `json:"layers,omitempty"`
	DryRun bool    `json:"dry_run,omitempty"`
	// Disk selects one owned disk by guest path; "/" selects the root. Conflicts with RootDiskOnly.
	Disk string `json:"disk,omitempty"`
	// RootDiskOnly excludes owned data disks. Conflicts with Disk.
	RootDiskOnly bool `json:"root_disk_only,omitempty"`
}

// DiskCompactionDiskResult reports one selected disk, including unchanged short chains.
type DiskCompactionDiskResult struct {
	GuestPath         string `json:"guest_path"`
	InputLayers       uint32 `json:"input_layers"`
	SelectedLayers    uint32 `json:"selected_layers"`
	OutputLayers      uint32 `json:"output_layers"`
	MaterializedBytes uint64 `json:"materialized_bytes"`
	// TotalUs measures preparation/materialization, excluding journal adoption and backend switching.
	TotalUs uint64 `json:"total_us"`
}

// DiskCompactionResult reports physical counts and measured durations in microseconds.
// MaterializedBytes is work performed, not an estimate of reclaimed disk space.
// TotalUs covers the whole operation, including shared journal/backend adoption.
type DiskCompactionResult struct {
	DryRun            bool                       `json:"dry_run"`
	InputLayers       uint32                     `json:"input_layers"`
	SelectedLayers    uint32                     `json:"selected_layers"`
	OutputLayers      uint32                     `json:"output_layers"`
	MaterializedBytes uint64                     `json:"materialized_bytes"`
	TotalUs           uint64                     `json:"total_us"`
	PauseUs           uint64                     `json:"pause_us"`
	Disks             []DiskCompactionDiskResult `json:"disks"`
}

// Compact merges selected sealed prefixes without rewriting existing snapshots.
// By default it covers the root and all sandbox-owned data disks, never named/external disks or directories.
func (s *Sandbox) Compact(ctx context.Context, opts DiskCompactionOptions) (*DiskCompactionResult, error) {
	data, err := json.Marshal(opts)
	if err != nil {
		return nil, err
	}
	out, err := s.inner.Compact(ctx, string(data))
	return parseCompaction(out, err)
}

// Compact performs explicit maintenance on a running or stopped sandbox.
func (h *SandboxHandle) Compact(ctx context.Context, opts DiskCompactionOptions) (*DiskCompactionResult, error) {
	// A metadata handle must not redirect a cloud sandbox's name into local storage.
	if h.backendKind != BackendLocal {
		return nil, &Error{Kind: ErrUnsupportedOperation, Message: "disk compaction requires a local sandbox handle"}
	}
	data, err := json.Marshal(opts)
	if err != nil {
		return nil, err
	}
	out, err := ffi.CompactSandbox(ctx, 0, h.name, string(data))
	return parseCompaction(out, err)
}

func parseCompaction(out string, err error) (*DiskCompactionResult, error) {
	if err != nil {
		return nil, wrapFFI(err)
	}
	var result DiskCompactionResult
	if err := json.Unmarshal([]byte(out), &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: sdk/node-ts/src/compact.ts
/** Explicit maintenance of the root and all owned data disks by default. */
export interface DiskCompactionOptions {
  /** Up to this many oldest sealed physical layers per disk, including the base; minimum two. */
  layers?: number;
  /** Only this owned disk's guest mount path; '/' selects the root. Conflicts with rootDiskOnly. */
  disk?: string;
  /** Only the root disk. Conflicts with disk. */
  rootDiskOnly?: boolean;
  dryRun?: boolean;
}

/** Per-disk physical counts; fewer than two selected layers leaves the disk unchanged. */
export interface DiskCompactionDiskResult {
  guestPath: string;
  inputLayers: number;
  selectedLayers: number;
  outputLayers: number;
  materializedBytes: number;
  /** Preparation/materialization time; excludes journal adoption and backend switching. */
  totalUs: number;
}

/** Timings are microseconds; materialized bytes are not reclaimed space. */
export interface DiskCompactionResult {
  dryRun: boolean;
  inputLayers: number;
  selectedLayers: number;
  outputLayers: number;
  materializedBytes: number;
  /** Whole-operation time, including the shared journal/backend adoption phase. */
  totalUs: number;
  pauseUs: number;
  disks: DiskCompactionDiskResult[];
}

export function compactionResultFromJson(json: string): DiskCompactionResult {
  const result = JSON.parse(json);
  return {
    dryRun: result.dry_run,
    inputLayers: result.input_layers,
    selectedLayers: result.selected_layers,
    outputLayers: result.output_layers,
    materializedBytes: result.materialized_bytes,
    totalUs: result.total_us,
    pauseUs: result.pause_us,
    disks: result.disks.map((disk: Record<string, number | string>) => ({
      guestPath: disk.guest_path as string,
      inputLayers: disk.input_layers as number,
      selectedLayers: disk.selected_layers as number,
      outputLayers: disk.output_layers as number,
      materializedBytes: disk.materialized_bytes as number,
      totalUs: disk.total_us as number,
    })),
  };
}
